crates/iceberg/src/arrow/caching_delete_file_loader.rs

```diff
@@ -0,0 +1,344 @@
+// under the License.
+
+use std::collections::HashMap;
+
+use futures::channel::oneshot;
+use futures::future::join_all;
+use futures::{StreamExt, TryStreamExt};
+use tokio::sync::oneshot::{channel, Receiver};
+
+use super::delete_filter::{DeleteFilter, EqDelFuture};
+use crate::arrow::delete_file_loader::BasicDeleteFileLoader;
+use crate::delete_vector::DeleteVector;
+use crate::expr::Predicate;
+use crate::io::FileIO;
+use crate::scan::{ArrowRecordBatchStream, FileScanTaskDeleteFile};
+use crate::spec::{DataContentType, SchemaRef};
+use crate::{Error, ErrorKind, Result};
+
+#[derive(Clone, Debug)]
+pub(crate) struct CachingDeleteFileLoader {
+    basic_delete_file_loader: BasicDeleteFileLoader,
+    concurrency_limit_data_files: usize,
+}
+
+// Intermediate context during processing of a delete file task.
+enum DeleteFileContext {
+    // TODO: Delete Vector loader from Puffin files
+    InProgEqDel(EqDelFuture),
+    PosDels(ArrowRecordBatchStream),
+    FreshEqDel {
+        batch_stream: ArrowRecordBatchStream,
+        sender: oneshot::Sender<Predicate>,
+    },
+}
+
+// Final result of the processing of a delete file task before
+// results are fully merged into the DeleteFileManager's state
+enum ParsedDeleteFileContext {
+    InProgEqDel(EqDelFuture),
+    DelVecs(HashMap<String, DeleteVector>),
+    EqDel,
+}
+
+#[allow(unused_variables)]
+impl CachingDeleteFileLoader {
+    pub(crate) fn new(file_io: FileIO, concurrency_limit_data_files: usize) -> Self {
+        CachingDeleteFileLoader {
+            basic_delete_file_loader: BasicDeleteFileLoader::new(file_io),
+            concurrency_limit_data_files,
+        }
+    }
+
+    /// Load the deletes for all the specified tasks
+    ///
+    /// Returned future completes once all loading has finished.
+    ///
+    ///  * Create a single stream of all delete file tasks irrespective of type,
+    ///      so that we can respect the combined concurrency limit
+    ///  * We then process each in two phases: load and parse.
+    ///  * for positional deletes the load phase instantiates an ArrowRecordBatchStream to
+    ///      stream the file contents out
+    ///  * for eq deletes, we first check if the EQ delete is already loaded or being loaded by
+    ///      another concurrently processing data file scan task. If it is, we return a future
+    ///      for the pre-existing task from the load phase. If not, we create such a future
+    ///      and store it in the state to prevent other data file tasks from starting to load
+    ///      the same equality delete file, and return a record batch stream from the load phase
+    ///      as per the other delete file types - only this time it is accompanied by a one-shot
+    ///      channel sender that we will eventually use to resolve the shared future that we stored
+    ///      in the state.
+    ///  * When this gets updated to add support for delete vectors, the load phase will return
```

Comment:
Looking forward to the puffin / deletion vector support!

Reply:
Me too! 😁